subxt/src/storage/storage_key.rs

```diff
@@ -0,0 +1,288 @@
+use crate::{
+    dynamic::DecodedValueThunk,
+    error::{Error, StorageAddressError},
+    metadata::{DecodeWithMetadata, Metadata},
+    utils::{Encoded, Static},
+};
+use scale_encode::EncodeAsType;
+use subxt_metadata::StorageHasher;
+
+use derivative::Derivative;
+
+/// This trait should be implemented by anything that can be used as one or multiple storage keys.
+pub trait StorageKey {
+    /// Iterator over the storage keys, each key implements EncodeAsType to
+    /// give the corresponding bytes to a `StorageHasher`.
+    fn keys_iter(&self) -> impl Iterator<Item = &dyn EncodeAsType>;
+
+    /// How many keys are there in total? Each key is an element that needs to be individually hashed.
+    // Note: Ideally we would use `impl ExactSizeIterator<Item = &dyn EncodeAsType>` for `keys_iter` above,
+    // But that plays poorly with the `Flatten` and `Chain` structs.
+    fn keys_len(&self) -> usize;
+
+    /// Attempts to decode the StorageKey from a storage address that has been stripped of its root bytes.
+    ///
+    /// Example: Imagine The `StorageKey` is a tuple (A,B) and the hashers are: [Blake2_128Concat, Twox64Concat].
+    /// Then the memory layout of the storage address is:
+    /// ```txt
+    /// | 8 bytes pallet hash | 8 bytes entry hash | 16 bytes hash of A | ... bytes of A | 8 bytes hash of B | ... bytes of B |
+    /// ```
+    /// `cursor` should point into a region after those first 16 bytes, at the start of a new hash.
+    /// `hashers_and_ty_ids` should consume all the hashers that have been used for decoding, such that there are less hashers coming to the next key.
+    fn decode_from_bytes(
+        cursor: &mut &[u8],
+        hashers_and_ty_ids: &mut &[(StorageHasher, u32)],
+        metadata: &Metadata,
+    ) -> Result<Self, Error>
+    where
+        Self: Sized + 'static;
+}
+
+/// Implement `StorageKey` for `()` which can be used for keyless storage entries.
+impl StorageKey for () {
+    fn keys_iter(&self) -> impl Iterator<Item = &dyn EncodeAsType> {
+        std::iter::empty()
+    }
+
+    fn keys_len(&self) -> usize {
+        0
+    }
+
+    fn decode_from_bytes(
+        _cursor: &mut &[u8],
+        hashers_and_ty_ids: &mut &[(StorageHasher, u32)],
+        _metadata: &Metadata,
+    ) -> Result<Self, Error> {
+        if hashers_and_ty_ids.is_empty() {
+            return Err(StorageAddressError::WrongNumberOfHashers {
+                hashers: 0,
+                fields: 1,
+            }
+            .into());
+        }
+        *hashers_and_ty_ids = &hashers_and_ty_ids[1..]; // Advance cursor by 1
+        Ok(())
+    }
+}
+
+/// A storage key for static encoded values.
+/// The original value is only present at construction, but can be decoded from the contained bytes.
+#[derive(Derivative)]
+#[derivative(Clone(bound = ""), Debug(bound = ""))]
+pub struct StaticStorageKey<K: ?Sized> {
+    pub(super) bytes: Static<Encoded>,
+    pub(super) _marker: std::marker::PhantomData<K>,
```

Comment:
Why `pub(super)` here? (Maybe it's because we construct this somewhere, and perhaps adding a constructor or `From` impl would avoid it? I'll probably spot why later in the review..)

Reply:
Oh this is not needed anymore at all, sorry.